### DISCUSSION TITLE
com.pi4j.exception.Pi4JException: Failed to execute action for device XXX on bus YYY

### DISCUSSION THREAD

**Discussion by razorne** [Pi4J/pi4j#276] [Q&A] [answered]
Hello,

My goal is to execute a blink using I2C and pi4j.

I checked multiple times the wiring and i can't find any error.

This is my code (mainly copied from pi4j docs):

```
import com.pi4j.Pi4J;
import com.pi4j.context.Context;
import com.pi4j.io.i2c.I2C;
import com.pi4j.io.i2c.I2CConfig;
import com.pi4j.io.i2c.I2CProvider;

public class MCP23017Reset {

    private static final int IODIRA = 0x00;
    private static final int IODIRB = 0x01;
    private static final int GPIOA = 0x12;
    private static final int GPIOB = 0x13;

    public static void main(String[] args) throws InterruptedException {
        reset(0x21);
    }

    protected static void reset(int address) throws InterruptedException {
        Context pi4j = Pi4J.newAutoContext();
        I2CProvider i2CProvider = pi4j.provider("linuxfs-i2c");
        I2CConfig i2cConfig = I2C.newConfigBuilder(pi4j).id("TCA9534").bus(1).device(address).build();

        try (I2C mcp23017Dev = i2CProvider.create(i2cConfig)) {
            System.out.println("i2c created");
            
            //configue all PinA output
            mcp23017Dev.writeRegister(IODIRA, (byte) 0x00);
            System.out.println("THIS LINE IS NEVER REACHED");
            
            //configue all PinB output
            mcp23017Dev.writeRegister(IODIRB, (byte) 0x00);
            System.out.println("register B is now output");

            mcp23017Dev.writeRegister(GPIOA, getAsInteger("11111111"));
            Thread.sleep(1000);
            mcp23017Dev.writeRegister(GPIOA, getAsInteger("00000000"));

            mcp23017Dev.writeRegister(GPIOB, getAsInteger("11111111"));
            Thread.sleep(1000);
            mcp23017Dev.writeRegister(GPIOB, getAsInteger("00000000"));
        }
    }

    protected static int getAsInteger(String binaryString) {
        return Integer.parseInt(binaryString, 2);
    }
}
```
The error i am getting is
```
Exception in thread "main" com.pi4j.exception.Pi4JException: Failed to execute action for device 33 on bus 1
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CBus.execute(LinuxFsI2CBus.java:95)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2C.write(LinuxFsI2C.java:82)
        at com.pi4j.io.IODataWriter.write(IODataWriter.java:106)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2C.writeRegister(LinuxFsI2C.java:118)
        at com.aconti.guitarrobot.main.MCP23017Reset.reset(MCP23017Reset.java:36)
        at com.aconti.guitarrobot.main.MCP23017Reset.main(MCP23017Reset.java:18)
Caused by: java.io.IOException: Remote I/O error
        at java.base/java.io.RandomAccessFile.writeBytes(Native Method)
        at java.base/java.io.RandomAccessFile.write(RandomAccessFile.java:559)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2C.lambda$write$1(LinuxFsI2C.java:83)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CBus.execute(LinuxFsI2CBus.java:80)
        ... 5 more
```

The line 'i2c created' is correctly printed out, the error comes up when i try to write to board register to configure it as output.

There is one weird doubt that i have... do i need a internet connection to run this? Because i have no internet now, whereas i had internet a couple of months ago when the program used to run correctly.

Thanks from my heart to anyone who will take their time to help me

**Comment by taartspi** [comment]
Hello
An internet connection is not required.
You should use the address identified with i2cdetect -y 1       0x20    This created your error.
Assuming the goal is drive an output pin High,You were missing setting the GPIOA bit for the pin you wanted to drive.
I connected an LED from pin0 to ground thru a resistor to prove  the code.
Also,  I connected my chip vcc to 3.3 to be compatible with the Pi GPIOs.
Also, tied the A0 A1 A2 to ground to be certain the chips address remained stable.

        public static void main(String[] args) throws InterruptedException {
            reset(0x20);
        }

        protected static void reset(int address) throws InterruptedException {
            Context pi4j = Pi4J.newAutoContext();
            I2CProvider i2CProvider = pi4j.provider("linuxfs-i2c");
            I2CConfig i2cConfig = I2C.newConfigBuilder(pi4j).id("TCA9534").bus(1).device(address).build();

            try (I2C mcp23017Dev = i2CProvider.create(i2cConfig)) {
                System.out.println("i2c created");


                //configue all PinA output
                mcp23017Dev.writeRegister(IODIRA, (byte) 0x00);
                System.out.println("GPIOA all pins configed as output");

                mcp23017Dev.writeRegister(GPIOA, getAsInteger("00000000"));
                System.out.println("GPIOA all pins set to 0");
                Thread.sleep(5000);

                //configue all PinB output
                mcp23017Dev.writeRegister(IODIRB, (byte) 0x00);
                System.out.println("register B is now output");


                mcp23017Dev.writeRegister(GPIOA, getAsInteger("00000001"));
                System.out.println("GPIOA pin 0 driven high, LED on");
                Thread.sleep(5000);
                mcp23017Dev.writeRegister(GPIOA, getAsInteger("00000000"));
                System.out.println("GPIOA pin 0 driven high, LED off");
                Thread.sleep(1000);


                mcp23017Dev.writeRegister(GPIOB, getAsInteger("00000001"));
                Thread.sleep(1000);
            }
        }

Let me know if this answers your questions or have additional questions.  Tom

**Comment by razorne** [comment]
Thanks a lot for your response Tom. I really appreciate it!

I tried your code but i am getting the same error. (I got error for device 32 instead of 33, since you used 0x20 instead of 0x21)

Please note that my code throws the error on this line 

`mcp23017Dev.writeRegister(IODIRA, (byte) 0x00);`

In fact this line `System.out.println("THIS LINE IS NEVER REACHED");` is never executed.

What this means is that i can successfully create the i2c bus, but as i try to set the registers as output, the error appears.

**Reply by taartspi** [reply]
I see I also tied the bar RESET to 3.3 v   Missed that in my first description

what address does i2cdetect -y 1 display ?

If the chip is in a bad state and no address is displayed, lift the and reattach VDD

**Reply by razorne** [reply]
I get 0x20 when i run that command

**Comment by taartspi** [comment]
Here is a README to document the mcp23017 pin connection, Pi snapshot and jdk
Also, your code with my changes to turn on and off an LED.  Note, had to change .java to .txt to attach the file.
[README.md](https://github.com/Pi4J/pi4j-v2/files/11413568/README.md)
[MCP23017_App.txt](https://github.com/Pi4J/pi4j-v2/files/11413570/MCP23017_App.txt)

**Reply by razorne** [reply]
Thanks for the time and effort you put in this answer.

As this solutions don't seem to work for my specific case, i assume i made a wiring error.

I will try to take some photos or draw some picture of the wiring and post those as well.

Again, thanks for your responses!!!

**Comment by razorne** [comment]
So... I made this test, i disconnected everything but the port expander (http://www.cqrobot.wiki/index.php/MCP23017_IO_Expansion_Board_SKU:_CQRMCP23017A) and raspberry. 
See picture
![IMG_6553](https://github.com/Pi4J/pi4j-v2/assets/57985543/cba2c691-c25e-4b2c-9d11-e3459401ca4b)
 I got the same error. How is this possible?

**Reply by taartspi** [reply, marked as answer]
I got a MCP23017 module like yours. 

CQRobot MCP23017

Pi SCL                      Blue
Pi SDA                      Green
Pi Gnd                      Black
Pi 3.3                        Red

PA0  ++++LED--/\/\/\/\/\/----  Gnd pin

I did no Ax jumpers  i2cdetect -y 1  shows I2C device address 0x27

Per CQRobot spec
/*the param can be 0 to 7,the default param is 7.means the dafault device address 0x27.

Test case executes and LED blinks on/off as expected.
I can only assume you have a Hardware Problem.  If other I2C devices function correctly I think your MCP23017 module is the problem.

**Reply by razorne** [reply]
You are right... OMG... default device il 0x27 when i thought it was 0x20. This is so embarassing... 

I tested this and it worked.

If i had internet i would had run that command to see that it was 0x27, but i have no cable connection right now... This also did hot help me in troubleshooting..

Again thank you very much for your time and effort to solve this!!!

**Reply by taartspi** [reply]
Very welcome

**Reply by razorne** [reply]
@taartspi Dear Sir, your help was really crucial, and i was hoping to ask you another question if you could be kind enough to answer it.
I have now created a successfull wiring between the above the device and raspberry.

The problem i am encountering now is with pi4j: i can't find a simple example that allows me to control the additional pins the port expander gives me. I saw you created this: 

https://github.com/Pi4J/pi4j-example-devices/tree/master/src/main/java/com/pi4j/devices/mcp23017 

Being a newbie in this field makes the approach to such documentation very hard.  I am very unsure about every step i make in the process of making the port expander work, because i really have a little, if no knowledge at all of the mechanics underneath.

Is there maybe an example like this one https://pi4j.com/documentation/io-examples/i2c/#code-example that could allow me to use your work?

I thank you very much for the effort you put into this!!!

**Reply by taartspi** [reply]
The base pi4j web pages won't contain device specific code.  It will show simple implementations of the various providers. There is the Examples pages  https://pi4j.com/examples/     My repo [https://github.com/Pi4J/pi4j-example-devices/tree/master/src/main/java/com/pi4j/devices also has a number of devices other than the MCP23017 . These example pages are not part of the pi4j directly, but offered as examples and when practicable usable code as is.     To continue development of code for the MCP23017 you would study the chips data sheet to understand the registers and their interaction.  There is a fair amount of function in this chip, walking through my code might assist in understanding those functions. You may not want to copy my directly but I think it will work as a starting point to learn from the datasheet.                                                                        All that said, I assume you have that simple program blinking the LED, if that is not true,  tell me the problems you encountered.  Tom                                                                                 Also I updated that little program to read a pin.
[README.md](https://github.com/Pi4J/pi4j-v2/files/12205705/README.md)   And again changed the .java to .txt as my session won't upload a java pgm         
[MCP23017_App.txt](https://github.com/Pi4J/pi4j-v2/files/12205738/MCP23017_App.txt)

**Reply by razorne** [reply]
Thank you very much for your response! My goal is kind of simple, i just want to control a large group (nearly 80) of pins from raspberry. All my pins are output. I need to send them 0s and 1s based on a specific schedule. At the end of the day my program is a big for loop: at every iteration, a bunch of 0s and 1s are sent to the MCP23017 devices. 

Something like this, you know:

```
for(int i=0;i<100;i++) {
  dev0x20.write(0x12, "10101111");
  dev0x20.write(0x13, "10101100");
  dev0x21.write(0x12, "10100111");
  dev0x21.write(0x13, "10101101");
  //other devices here...
  Thread.sleep(1000L);
}
```
I tried to use the MCP23017 device bought from CQRobot with this code example https://pi4j.com/documentation/io-examples/i2c/#code-example , but it just doesn't work properly. With this code example i can target 1 single pin correctly. But if i try to target two or more pins (for example sending 11000000) all the pins are targeted simultaneously, just like i sent 11111111 or 00000000. And this makes sense, for a newbie like me, because that code example is meant for a TCA9534 chip

```

**Reply by razorne** [reply]
Since you bought the same chip. Could you confirm that you get the same behaviour (i.e when you target 2 or more pins, ALL pins are targeted?)

**Comment by taartspi** [comment]
Data    "10101111"   would then be this byte array being sent to the mcp23017   49 48 49 48 49 49 49 49   The chip expects a single byte of data.   At the moment I have a different 23017 plugged in but it appears to ignore all the bytes and not drive any pins. 

As an example of driving GPA 0 and 1,        mcp23017Dev.writeRegister(GPIOA, 0b00000011);

**Reply by razorne** [reply]
Still :( i tried your suggestion but all pins are being targeted

This is the code i run:

```
import com.pi4j.Pi4J;
import com.pi4j.context.Context;
import com.pi4j.io.i2c.I2C;
import com.pi4j.io.i2c.I2CConfig;
import com.pi4j.io.i2c.I2CProvider;
import java.io.IOException;

public class I2CTesterBy2 {

    private static final int IODIRA = 0x00;
    private static final int GPIOA = 0x12;
    private static final int IODIRB = 0x01;

    public static void main(String[] args) throws Exception {
        blink(32);
    }

    protected static void blink(int address) throws InterruptedException, IOException {
        Context pi4j = Pi4J.newAutoContext();

        I2CProvider i2CProvider = pi4j.provider("linuxfs-i2c");
        I2CConfig i2cConfig = I2C.newConfigBuilder(pi4j).id("TCA9534").bus(1).device(address).build();

        try (I2C mcp23017Dev = i2CProvider.create(i2cConfig)) {
            System.out.println("I2C is created");
            mcp23017Dev.writeRegister(IODIRA, (byte) 0x00);
            System.out.println("Register A is configured as output");
            mcp23017Dev.writeRegister(IODIRB, (byte) 0x00);
            System.out.println("Register B is configured as output");

            Thread.sleep(1000);

            for (int i = 0; i < 10; i++) {
                mcp23017Dev.writeRegister(GPIOA, 0b00000011);
                Thread.sleep(1000);
                mcp23017Dev.writeRegister(GPIOA, 0x00);
                Thread.sleep(1000);
            }
        }
    }
}
```

**Reply by taartspi** [reply]
I dug out my ocean  mcp23017 and wired it as listed in the readme.  I did connect a second LED to PA01           I copied your code into a new file and only changed the device address to 0x27, you had been using 32 (0x20)     Your code function as you want and the pair of LEDs blink ten time.   Maybe double check the resistor used with the LED and the polarity of your connection.       To validate the chip functions you could use the last version of my program to check using a pin as an input.

**Reply by razorne** [reply]
Hello and thank you very much for your reply and your support! I am glad your are able to make it work, and also that the code works as expected. This is good news. There must be some problem with my wiring then.

But the device 0x20 is correct as i am getting this:

![image](https://github.com/Pi4J/pi4j-v2/assets/57985543/0bda2054-0096-474a-9c82-25d6ef453c91)

I am trying to control a bunch of relays as you can see from the picture (so there is no led resistor). I can of course prepare a LED sample with my breadboard. I will be able to make some tests tomorrow and let you know.

![IMG_7035](https://github.com/Pi4J/pi4j-v2/assets/57985543/df3a8f63-fd88-4ff1-92d7-7ddc9542ef89)

I have a question thought: 1) So Pi4J supports MCP23017? Is this piece of code is irrelevant `(pi4j).id("TCA9534")` ? 2) I don't think i correctly understand what should i put here mcp23017Dev.writeRegister(GPIOA, WHAT_EXACTLY_GOES_HERE?); At first i thought that i could pass a string of 8 characters, 0s or 1s, and a single character indicated low or high for a specific pin on the register. But based on your answer i got it wrong

**Reply by taartspi** [reply]
Yes i agree to connect a simple led and verify operation.    I showed the mcp23017 connected to 3.3 v for pi gpio compatibility. Are your trigger devices usable with 3.3 or do they expect 5v?  Does the mcp supply sufficient current to control the triggers ?      You need that statement , make the id string tca9548 something meaningful to your work.   Write register has many signatures.  The mcp chip expects one byte read and writes.

**Reply by razorne** [reply]
1) Ok i will make a test with breadboard and LEDs;
2) Relays are powered up by external 12v transformer. I would be sure that mcp can trigger them, since as of today i can trigger all pins or 1 pin at a time;
3) You need that statement , make the id string tca9548 something meaningful to your work --> I am not sure what you mean by that
4) I will look into this!

Thanks!!!

**Reply by razorne** [reply]
Update:

I tried to execute the code on a different MCP230217 device with relays and the outcome did not change (all up/down... could not control just the first 2 pins). So the problem does not lie with the device.

I will later try to create a LEDs circuit...

**Reply by razorne** [reply]
I was unable to replicate this with LEDs... None of my LEDs blinked.

LED anode is supposed to connected to PA0, right?
Whereas LED cathode is supposed to be connected to negative coming from the device, right?

![IMG_7099](https://github.com/Pi4J/pi4j-v2/assets/57985543/cbac0b3d-d9df-4a50-a8c9-5b7b04881b3b)
![IMG_7100](https://github.com/Pi4J/pi4j-v2/assets/57985543/f718d765-f42c-4cd2-a2f7-2a7d00b27c70)


![IMG_7101](https://github.com/Pi4J/pi4j-v2/assets/57985543/36f232ca-71b2-426e-8e28-d9aa841a9188)
![IMG_7102](https://github.com/Pi4J/pi4j-v2/assets/57985543/a7f49b0b-a1b3-4fa0-aabd-7dc98000391e)

**Reply by taartspi** [reply]
Yes it is PA0     I would disconnect every thing on the MCP except the white connector.  Check the Pi connections from the white connector.   i2cdetect command to ensure the chip is visible to the software.    hook up the LEDs  Make sure the negative LED leg is the one connected to the negative (-) of the breadboard.  Connect the the negative lead from the bread board to the GND pin on the MCP,  Connect the positive lead from the breadboard to the PA0.    The other LED lead to PA01.     Execute that program you sent to a couple days ago.   Assuming that works begin to restore your other connections until the test fails.

**Comment by razorne** [comment]
Ok so... with the LEDs, unfortunately... it properly works... i can control groups of pins, not just all or none.

So, what does this mean? What is the problem with controlling relays instead of LEDs?

**Reply by taartspi** [reply]
As I mentioned earlier "I showed the mcp23017 connected to 3.3 v for pi gpio compatibility. Are your trigger devices usable with 3.3 or do they expect 5v? Does the mcp supply sufficient current to control the triggers ?"     My debug path would be to connect a single 'relay' to the MCP23017/Pi, recheck the wiring and switches.  If it still fails I would 'google' to find others that encountered similar problems and how they fixed the problem.  And possibly the 'relay' manufacturer has a forum where you can ask questions.

**Comment by razorne** [comment]
I finally solved the issue. The problem lied with the wiring of course... basically the reference to 0 was not shared amongst the devices and that caused an unpredictable, undesired, outcome...

On the other hand the Pi4J works smoothly with MCP23017. I am able to write to my register like so
device.writeRegister(0x12, Integer.parseInt("10000000", 2));

**Reply by FDelporte** [reply]
thanks @razorne for the feedback and @taartspi for the help!

in the past I also made the mistake a few times of not connecting RPi and components to the same ground. Looks like something we need to document somewhere...

https://github.com/Pi4J/pi4j.github.io/issues/42
